File: iosApp/iosApp/Features/LocalOffices/LocalOfficeLinkEditorSheet.swift
import SwiftUI
import Shared

// MARK: - Editor mode

enum LinkEditorMode {
    case createSection
    case createLink
    case edit(existing: LocalOfficeLinktreeRowModel)
}

// MARK: - Sheet

struct LocalOfficeLinkEditorSheet: View {
    let officeId: String
    /// All sibling entries sharing the same parent — used to auto-compute sort_order for new entries.
    let siblings: [LocalOfficeLinktreeRowModel]
    let parentCandidates: [LocalOfficeLinktreeRowModel]  // already filtered to kind == "section"
    let mode: LinkEditorMode

    // Form state
    @State private var kind: String          // "section" or "link"
    @State private var title: String
    @State private var url: String
    @State private var icon: String
    @State private var parentId: String      // "" = root
    @State private var sortOrder: Int        // computed, not exposed in UI
    @State private var active: Bool

    // UI state
    @State private var saving = false
    @State private var errorMessage: String? = nil
    @State private var showError = false

    @Environment(\.dismiss) private var dismiss

    init(officeId: String, siblings: [LocalOfficeLinktreeRowModel], parentCandidates: [LocalOfficeLinktreeRowModel], mode: LinkEditorMode) {
        self.officeId = officeId
        self.siblings = siblings
        self.mode = mode

        // Auto-compute sort_order = (max existing in same parent) + 1 for new entries.
        let maxOrder = siblings.map { Int($0.sortOrder) }.max() ?? 0

        switch mode {
        case .createSection:
            _kind = State(initialValue: "section")
            _title = State(initialValue: "")
            _url = State(initialValue: "")
            _icon = State(initialValue: "")
            _parentId = State(initialValue: "")
            _sortOrder = State(initialValue: maxOrder + 1)
            _active = State(initialValue: true)
            self.parentCandidates = parentCandidates
        case .createLink:
            _kind = State(initialValue: "link")
            _title = State(initialValue: "")
            _url = State(initialValue: "")
            _icon = State(initialValue: "")
            _parentId = State(initialValue: "")
            _sortOrder = State(initialValue: maxOrder + 1)
            _active = State(initialValue: true)
            self.parentCandidates = parentCandidates
        case .edit(let existing):
            _kind = State(initialValue: existing.kind)
            _title = State(initialValue: existing.title)
            _url = State(initialValue: existing.url)
            _icon = State(initialValue: existing.icon)
            _parentId = State(initialValue: existing.parent)
            _sortOrder = State(initialValue: Int(existing.sortOrder))
            _active = State(initialValue: true)   // view model doesn't expose active; default true
            // Exclude self from parent candidates (can't be its own parent)
            self.parentCandidates = parentCandidates.filter { $0.id != existing.id }
        }
    }

    var body: some View {
        NavigationStack {
            Form {
                // Kind picker
                Section {
                    Picker(selection: $kind) {
                        Text(tr("local_office.editor.kind.section", fallback: "Sezione"))
                            .tag("section")
                        Text(tr("local_office.editor.kind.link", fallback: "Link"))
                            .tag("link")
                    } label: {
                        EmptyView()
                    }
                    .pickerStyle(.segmented)
                    .listRowInsets(EdgeInsets(top: 8, leading: 16, bottom: 8, trailing: 16))
                }

                // Title (required)
                Section {
                    TextField(
                        tr("local_office.editor.title", fallback: "Titolo"),
                        text: $title
                    )
                }

                // URL — only when kind == link
                if kind == "link" {
                    Section {
                        TextField(
                            tr("local_office.editor.url", fallback: "URL"),
                            text: $url
                        )
                        .keyboardType(.URL)
                        .autocorrectionDisabled()
                        .textInputAutocapitalization(.never)
                    }
                }

                // Icon
                Section {
                    TextField(
                        tr("local_office.editor.icon", fallback: "Icona"),
                        text: $icon
                    )
                    Text(tr("local_office.editor.icon.hint", fallback: "Inserisci un emoji o il nome di un SF Symbol"))
                        .font(.caption)
                        .foregroundStyle(.secondary)
                }

                // Parent picker — show when there are candidate sections
                if !parentCandidates.isEmpty {
                    Section(tr("local_office.editor.parent", fallback: "Sezione padre")) {
                        Picker(selection: $parentId) {
                            Text(tr("local_office.editor.parent.none", fallback: "Nessuna (root)"))
                                .tag("")
                            ForEach(parentCandidates, id: \.id) { section in
                                Text(section.title).tag(section.id)
                            }
                        } label: {
                            EmptyView()
                        }
                        .pickerStyle(.inline)
                        .labelsHidden()
                    }
                }

                // Active toggle
                Section {
                    Toggle(tr("local_office.editor.active", fallback: "Attivo"), isOn: $active)
                }
            }
            .navigationTitle(navTitle)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button(tr("local_office.editor.cancel", fallback: "Annulla")) {
                        dismiss()
                    }
                }
                ToolbarItem(placement: .confirmationAction) {
                    if saving {
                        ProgressView()
                    } else {
                        Button(tr("local_office.editor.save", fallback: "Salva")) {
                            Task { await save() }
                        }
                        .fontWeight(.semibold)
                        .disabled(title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
                    }
                }
            }
            .alert(
                tr("app.error.title", fallback: "Errore"),
                isPresented: $showError,
                presenting: errorMessage
            ) { _ in
                Button("OK", role: .cancel) {}
            } message: { msg in
                Text(msg)
            }
        }
    }

    // MARK: - Helpers

    private var navTitle: String {
        switch mode {
        case .createSection:
            return tr("local_office.links.add_section", fallback: "Aggiungi sezione")
        case .createLink:
            return tr("local_office.links.add_link", fallback: "Aggiungi link")
        case .edit:
            return tr("local_office.links.edit", fallback: "Modifica voce")
        }
    }

    private var isCreating: Bool {
        switch mode {
        case .createSection, .createLink: return true
        case .edit: return false
        }
    }

    private func save() async {
        let trimmedTitle = title.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedTitle.isEmpty else { return }

        saving = true
        defer { saving = false }

        do {
            switch mode {
            case .createSection, .createLink:
                // Use the field-by-field create wrapper. Constructing the
                // `LocalOfficeLinkRecord` from Swift with a Boolean field
                // dropped the value across the K/N bridge (active landed as
                // `false` server-side even when the Toggle was `true`).
                _ = try await koin.localOffices.createLinkFromFields(
                    officeId: officeId,
                    kind: kind,
                    parent: parentId,
                    title: trimmedTitle,
                    url: url.trimmingCharacters(in: .whitespacesAndNewlines),
                    icon: icon.trimmingCharacters(in: .whitespacesAndNewlines),
                    sortOrder: Int32(sortOrder),
                    active: active
                )

            case .edit(let existing):
                _ = try await koin.localOffices.updateLinkFields(
                    officeId: officeId,
                    id: existing.id,
                    kind: kind,
                    parent: parentId,
                    title: trimmedTitle,
                    url: url.trimmingCharacters(in: .whitespacesAndNewlines),
                    icon: icon.trimmingCharacters(in: .whitespacesAndNewlines),
                    sortOrder: KotlinInt(integerLiteral: sortOrder),
                    active: KotlinBoolean(bool: active)
                )
            }
            dismiss()
        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }
    }
}

File: androidApp/src/main/kotlin/it/mensa/app/features/localoffices/_components/LocalOfficeLinkEditorSheet.kt
package it.mensa.app.features.localoffices._components

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Save
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.ui.unit.dp
import it.mensa.app.support.koinAccess
import it.mensa.app.support.tr
import it.mensa.shared.model.LocalOfficeLinktreeRowModel
import kotlinx.coroutines.launch

sealed class LinkEditorMode {
    object CreateSection : LinkEditorMode()
    object CreateLink : LinkEditorMode()
    data class Edit(val existing: LocalOfficeLinktreeRowModel) : LinkEditorMode()
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun LocalOfficeLinkEditorSheet(
    officeId: String,
    siblings: List<LocalOfficeLinktreeRowModel>,
    parentCandidates: List<LocalOfficeLinktreeRowModel>, // kind == "section"
    mode: LinkEditorMode,
    onDismiss: () -> Unit,
) {
    val repo = remember { koinAccess().localOffices }
    val scope = rememberCoroutineScope()

    val maxOrder = siblings.maxOfOrNull { it.sortOrder } ?: 0

    var kind by remember {
        mutableStateOf(
            when (mode) {
                is LinkEditorMode.CreateSection -> "section"
                is LinkEditorMode.CreateLink -> "link"
                is LinkEditorMode.Edit -> mode.existing.kind
            }
        )
    }
    var title by remember {
        mutableStateOf(if (mode is LinkEditorMode.Edit) mode.existing.title else "")
    }
    var url by remember {
        mutableStateOf(if (mode is LinkEditorMode.Edit) mode.existing.url else "")
    }
    var icon by remember {
        mutableStateOf(if (mode is LinkEditorMode.Edit) mode.existing.icon else "")
    }
    var parentId by remember {
        mutableStateOf(if (mode is LinkEditorMode.Edit) mode.existing.parent else "")
    }
    var active by remember {
        mutableStateOf(true)
    }
    var saving by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    val isCreating = mode !is LinkEditorMode.Edit
    val navTitle = when (mode) {
        is LinkEditorMode.CreateSection -> tr("local_office.links.add_section", fallback = "Aggiungi sezione")
        is LinkEditorMode.CreateLink -> tr("local_office.links.add_link", fallback = "Aggiungi link")
        is LinkEditorMode.Edit -> tr("local_office.links.edit", fallback = "Modifica voce")
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 16.dp)
                .padding(bottom = 32.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            // Header
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                TextButton(onClick = onDismiss) {
                    Text(tr("local_office.editor.cancel", fallback = "Annulla"))
                }
                Text(navTitle, style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
                if (saving) {
                    CircularProgressIndicator(modifier = Modifier.size(24.dp))
                } else {
                    TextButton(
                        onClick = {
                            val trimTitle = title.trim()
                            if (trimTitle.isEmpty()) return@TextButton
                            saving = true
                            scope.launch {
                                try {
                                    if (isCreating) {
                                        repo.createLinkFromFields(
                                            officeId = officeId,
                                            kind = kind,
                                            parent = parentId,
                                            title = trimTitle,
                                            url = url.trim(),
                                            icon = icon.trim(),
                                            sortOrder = maxOrder + 1,
                                            active = active,
                                        )
                                    } else {
                                        val existing = (mode as LinkEditorMode.Edit).existing
                                        repo.updateLinkFields(
                                            officeId = officeId,
                                            id = existing.id,
                                            kind = kind,
                                            parent = parentId,
                                            title = trimTitle,
                                            url = url.trim(),
                                            icon = icon.trim(),
                                            active = active,
                                        )
                                    }
                                    onDismiss()
                                } catch (e: Exception) {
                                    error = e.message
                                } finally {
                                    saving = false
                                }
                            }
                        },
                        enabled = title.trim().isNotEmpty(),
                    ) {
                        Text(
                            tr("local_office.editor.save", fallback = "Salva"),
                            fontWeight = FontWeight.SemiBold,
                        )
                    }
                }
            }

            // Kind segmented picker
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                listOf("section" to tr("local_office.editor.kind.section", fallback = "Sezione"),
                    "link" to tr("local_office.editor.kind.link", fallback = "Link")).forEach { (k, label) ->
                    FilterChip(
                        selected = kind == k,
                        onClick = { kind = k },
                        label = { Text(label) },
                        modifier = Modifier.weight(1f),
                    )
                }
            }

            // Title
            OutlinedTextField(
                value = title,
                onValueChange = { title = it },
                label = { Text(tr("local_office.editor.title", fallback = "Titolo")) },
                modifier = Modifier.fillMaxWidth(),
                singleLine = true,
            )

            // URL (only for links)
            if (kind == "link") {
                OutlinedTextField(
                    value = url,
                    onValueChange = { url = it },
                    label = { Text(tr("local_office.editor.url", fallback = "URL")) },
                    modifier = Modifier.fillMaxWidth(),
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Uri),
                )
            }

            // Icon
            Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                OutlinedTextField(
                    value = icon,
                    onValueChange = { icon = it },
                    label = { Text(tr("local_office.editor.icon", fallback = "Icona")) },
                    modifier = Modifier.fillMaxWidth(),
                    singleLine = true,
                )
                Text(
                    tr("local_office.editor.icon.hint", fallback = "Inserisci un emoji o il nome di un SF Symbol"),
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            // Parent picker
            if (parentCandidates.isNotEmpty()) {
                Text(
                    tr("local_office.editor.parent", fallback = "Sezione padre"),
                    style = MaterialTheme.typography.labelLarge,
                )
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable { parentId = "" },
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    RadioButton(selected = parentId.isEmpty(), onClick = { parentId = "" })
                    Text(tr("local_office.editor.parent.none", fallback = "Nessuna (root)"))
                }
                parentCandidates.forEach { sec ->
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .clickable { parentId = sec.id },
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        RadioButton(selected = parentId == sec.id, onClick = { parentId = sec.id })
                        Text(sec.title)
                    }
                }
            }

            // Active toggle
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(tr("local_office.editor.active", fallback = "Attivo"))
                Switch(checked = active, onCheckedChange = { active = it })
            }
        }
    }

    error?.let { err ->
        AlertDialog(
            onDismissRequest = { error = null },
            title = { Text(tr("app.error.title", fallback = "Errore")) },
            text = { Text(err) },
            confirmButton = { TextButton(onClick = { error = null }) { Text("OK") } },
        )
    }
}
